### ISSUE TITLE
train_hybrid.py: per-head scalar normalization — MLP loss is dominated by widest-range output, narrower heads can't fit

### ISSUE THREAD

**Issue by lilith** [imazen/zenanalyze#67]
## TL;DR

`train_hybrid.py` concatenates `bytes_log` + N scalar prediction heads into one wide soft-target vector and fits an MLP with vanilla MSE loss. Head-loss is unweighted and unnormalized, so the student MLP is dominated by whichever head has the widest output range. **Narrower-range heads collapse with student-RMSE >> teacher-RMSE.**

## Reproduction (zenwebp v0.2, 2026-05-01)

Added `multi_pass_stats` (binary 0/1) as a 5th scalar head alongside the existing 0..100 / 0..7 ranged heads. Trainer ran cleanly. Result:

```
Teacher multi_pass_stats RMSE: 0.26    (reasonable for binary)
Student multi_pass_stats RMSE: 27.7    ← 100× the data range

Teacher filter_sharpness RMSE: 2.18    (range 0..7)
Student filter_sharpness RMSE: 40.7    ← also broken

Teacher sns_strength RMSE:    33.5
Student sns_strength RMSE:    52.6     ← regressed

Student argmin mean overhead: 3.55%   (was 1.99% on v0.1 with 3 heads)
Student argmin acc:           16.6%   (was 62.1% on v0.1)
```

The model isn't broken; the loss function is. The student got pulled toward fitting `bytes_log` (~range 10) and the wider scalars while the narrow ones drifted. Once we drop the binary head, the wider heads also recover (retrain currently in progress).

## Root cause in `train_hybrid.py`

Around line 950 (post-2026-04-29 rename, in `zentrain/tools/`):

```python
soft_tr = np.concatenate(
    [bytes_pred_tr] + [scalar_pred_tr[axis] for axis in SCALAR_AXES], axis=1
)
# ...
student.fit(Xe_tr_s, soft_tr)
```

Both sklearn's `MLPRegressor` and the pytorch `LeakyReLU` backend in `_train_torch_leakyrelu_student` compute MSE over the concatenated output. With heads of widely different ranges:

- `bytes_log` (range ~5..15)
- `chroma_scale` (0.6..1.5)
- `lambda` (8..25)
- `sns_strength` (0..100)
- `filter_strength` (0..100)
- `filter_sharpness` (0..7)
- `partition_limit` (0..100)
- `multi_pass_stats` (0..1)

…the loss landscape is dominated by the wide heads. Narrow heads contribute negligibly to the gradient.

## Proposed fix — per-head min-max normalization

1. **Trainer:** before `student.fit`, compute `(lo, hi)` per output column on the training soft-targets. Scale each column to `[0, 1]`. Store the array of `(lo, hi)` pairs in the model JSON.
2. **Bake:** pass the array through `bake_picker.py` → ZNPR v2 metadata as a typed numeric TLV entry.
3. **Runtime:** `Predictor::predict` returns scaled `[0, 1]` outputs; codec consumers denormalize per-head before consuming. `argmin_masked_in_range` over the bytes_log block continues to work — argmin in linear-bytes space is invariant to per-column monotonic min-max.

Three load-bearing properties of this approach:

- **`bytes_log` argmin invariant.** Min-max preserves order. The codec's `argmin_masked_in_range(.., bytes_log_range, .., ScoreTransform::Exp, ..)` keeps working without changes — it's an order operation, not a value operation.
- **Scalar denormalization is simple.** Per-cell scalar reads become `lo[axis] + (hi[axis] - lo[axis]) * predicted`. Constant-time, no model changes.
- **Bake format compat.** ZNPR v2's metadata blob already carries arbitrary numeric arrays via `MetadataEntryJson::f32`. Add a `zenpicker.head_norm_lo` / `zenpicker.head_norm_hi` pair; runtime reads at startup.

## Out-of-scope

- Sample-weighted multi-output MSE (would require pytorch-only path; sklearn's MLPRegressor doesn't support per-output weights). Min-max approach works on both backends.
- Z-score normalization. Tested mentally; works equivalently for argmin on bytes_log; adds slight complexity for scalars (you'd want both mean and std). Min-max is cleaner.

## Why this matters

Every codec adopting hybrid heads will eventually want a binary or narrow-range scalar head:
- zenjpeg's `lambda` already has effectively-binary "trellis on/off" semantics via the sentinel
- zenavif will likely want `tile` / `still picture` flags
- zenjxl will want effort sub-flags

Without normalization, every codec re-discovers the same loss-imbalance regression we just hit on zenwebp v0.2.

## Test plan

- [ ] Trainer emits `head_norm_lo` / `head_norm_hi` in model JSON
- [ ] `bake_picker.py` passes them through to v2 metadata
- [ ] zenjpeg v2.2 trains bit-exactly when the new normalization stage is no-op (i.e., already-normalized to [0,1] range)
- [ ] zenwebp v0.2 with all 5 scalar heads trains without student-RMSE explosion on `multi_pass_stats`
- [ ] `Predictor::predict` documentation explains the output is scaled; codec runtime denormalizes

I'd take this on after the current zenwebp v0.2 work lands on a workable picker (which can ship with 4 scalars and no multi_pass_stats while this fix is in flight).

**Comment by lilith** [comment]
Closing — verified stale/resolved on the current trunk.

**Summary:** Fixed by 7e2534a: train_hybrid.py standardizes each scalar head before fitting and folds the inverse affine into the final layer (train_hybrid.py:3650ff, σ≈0 guards at :3693/:3701), so narrow-range heads get balanced gradient without any bake-format or runtime change.

**Evidence (checked against `origin` trunk on 2026-08-27/28):**
7e2534a 'feat(zentrain): per-head loss normalization in train_hybrid.py' is an ancestor of origin/main. train_hybrid.py:3650 per-head loss normalization block, :3693/:3701 σ≈0 guards; :266/:288/:3181 reference the pre-#67 behaviour. Design differs from the proposal: scalar blocks are standardized before fit and the inverse affine is folded into the final layer (commit body), so predictions stay in natural units, ZNPR is unchanged, and no head_norm_lo/hi metadata or runtime denormalization is needed — test-plan items 1/2/5 are moot by construction. Commit body records pre-fix 581%/2770% normalized RMSE on zenavif/zenwebp; a post-fix zenwebp 5-head number is not in the commit body (not verified here).

**References:** `7e2534a`; files: `train_hybrid.py:3650`

---
*This issue was reviewed and closed by Claude (Anthropic) operating under the repo owner's account, as part of the 2026-08-26/27 cross-repo issue triage: a single-pass review classified it as likely stale, and an independent read-only verification then tried to refute that against the trunk and could not. If any part of the original ask is still open, reopen with the specific item — nothing in this comment was generated without opening the cited files.*
